### PR TITLE
Refactor out ping handling

### PR DESCRIPTION
**Pull request by jakhog**
We have already merged the interfaces so we could continue other work - sorry about that. Bat all the files have been changed here, so if you just look at the files (and ignore what GitHub says has been changed) you'll see what is new.

There are a few pieces still missing for this to work properly, that we can do later:
- Implement the `IKeepConnectionsAlive` interface (which is just a factory that will take some dependencies and pass them to the pinged connections).
- Implement and bind the `IMetricsCollector` (we can do a null implementation for now).

I purposefully did not bind the `TokenSourceDeadline` class to the interface, as I seem to recall we have issues with DI and disposables?

Also, a lot of specs for PingedConnection still missing - but it makes more sense to do that once the callback timer interface is ready.

### CODE REVIEW

File: Source/Services/ReverseCalls/TokenSourceDeadline.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (_disposed) return;
+
+            if (disposing)
+            {
+                _source.Dispose();
+            }
```
Comment (woksin) [inline Source/Services/ReverseCalls/TokenSourceDeadline.cs:46]: Not cancel? See previous comment about dispose pattern

File: Source/Services/ReverseCalls/WrappedAsyncStreamWriter.cs [hunk 2]
```diff
@@ -68,37 +77,81 @@ public WriteOptions WriteOptions
         /// <returns>A task that will complete when the message is sent.</returns>
         public async Task WriteAsync(TServerMessage message)
         {
+            _logger.WritingMessage(_requestId, typeof(TServerMessage));
             _cancellationToken.ThrowIfCancellationRequested();
+
             if (!_writeLock.Wait(0))
```
Comment (woksin) [inline Source/Services/ReverseCalls/WrappedAsyncStreamWriter.cs:83]: aren't there async locks? I think there might be one in the Nito.AsyncEx packages I'm using in EventHorizon

File: Source/Services/ReverseCalls/WrappedAsyncStreamWriter.cs [hunk 3]
```diff
@@ -68,37 +77,81 @@ public WriteOptions WriteOptions
...
+
                 stopwatch.Stop();
+                _metrics.DecrementPendingStreamWrites();
                 _metrics.AddToTotalStreamWriteWaitTime(stopwatch.Elapsed);
+                _logger.WritingMessageUnblockedAfter(_requestId, typeof(TServerMessage), stopwatch.Elapsed);
             }
 
-            await WriteRecordMetricsAndReleaseLock(message, false);
+            try
+            {
```
Comment (woksin) [inline Source/Services/ReverseCalls/WrappedAsyncStreamWriter.cs:99]: This should perhaps be moved right under the acquiring of the lock

File: Source/Services/ReverseCalls/WrappedAsyncStreamWriter.cs (hunk 2 above)
Comment (woksin) [reply Source/Services/ReverseCalls/WrappedAsyncStreamWriter.cs:83]: Hmmm, I might just don't know the API well enough :P

**Review by woksin — APPROVED** [review]

**Review by woksin — APPROVED** [review]

File: Source/Services/ReverseCalls/PingedConnection.cs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+                _cancellationTokenSource.Cancel();
+                WaitForPingStarterToCompleteIgnoringExceptions();
+                MaybeStopPinging();
+                _keepAliveExpiredRegistration.Dispose();
+                _keepalive.Dispose();
+                _wrappedReader.MessageReceived -= ResetKeepaliveTokenCancellation;
+                _cancellationTokenSource.Dispose();
+                _wrappedWriter.Dispose();
+                _logger.DisposedPingedConnection(_requestId);
+            }
```
[the remark this answers is not in the document]
Comment (jakhog) [reply Source/Services/ReverseCalls/PingedConnection.cs:124]: We discussed this - and agreed it was correct as it is

File: Source/Services/ReverseCalls/TokenSourceDeadline.cs (hunk 1 above)
Comment (jakhog) [reply Source/Services/ReverseCalls/TokenSourceDeadline.cs:46]: Same here for dispose.

File: Source/Services/ReverseCalls/PingedConnection.cs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+            _metrics.IncrementTotalKeepaliveTimeouts();
+            _logger.KeepaliveTimedOut(_requestId);
+        }
+
+        void WaitForPingStarterToCompleteIgnoringExceptions()
+        {
+            var stopwatch = Stopwatch.StartNew();
+            _logger.WaitingForPingStarterToComplete(_requestId);
+
+            _pingStarter.ContinueWith(_ => {}).GetAwaiter().GetResult();
```
[the remark this answers is not in the document]
Comment (jakhog) [reply Source/Services/ReverseCalls/PingedConnection.cs:193]: Maybe not clear enough by the method name, but it is a simple way to ignore exceptions and cancellations without having `GetResult()` throw an exception that we need to catch.

File: Source/Services/ReverseCalls/WrappedAsyncStreamWriter.cs (hunk 2 above)
Comment (jakhog) [reply Source/Services/ReverseCalls/WrappedAsyncStreamWriter.cs:83]: In this case the `.Wait(0)` is a non-blocking attempt to take the lock. If it fails, it uses the lock asynchronously below.

File: Source/Services/ReverseCalls/WrappedAsyncStreamWriter.cs (hunk 3 above)
Comment (jakhog) [reply Source/Services/ReverseCalls/WrappedAsyncStreamWriter.cs:99]: Will fix and make another PR

**Pull request merged**